crates/bevy_pbr/src/material.rs

```diff
@@ -210,11 +209,27 @@ impl<M: SpecializedMaterial> Plugin for MaterialPlugin<M> {
                 .add_render_command::<AlphaMask3d, DrawMaterial<M>>()
                 .init_resource::<MaterialPipeline<M>>()
                 .init_resource::<SpecializedMeshPipelines<MaterialPipeline<M>>>()
+                .add_system_to_stage(RenderStage::Extract, extract_materials::<M>)
                 .add_system_to_stage(RenderStage::Queue, queue_material_meshes::<M>);
         }
     }
 }
 
+fn extract_materials<M: SpecializedMaterial>(
+    mut commands: Commands,
+    query: Query<(Entity, &ComputedVisibility, &Handle<M>)>,
+    mut prev_len: Local<usize>,
+) {
+    let mut materials = Vec::with_capacity(*prev_len);
+    for (entity, computed_visibility, material) in query.iter() {
+        if computed_visibility.is_visible {
+            materials.push((entity, (material.clone(),)));
+        }
+    }
+    *prev_len = materials.len();
+    commands.insert_or_spawn_batch(materials);
```

Comment:
insert_or_spawn_batch takes an iterator. You could probably just take query.iter(), filter it, map it, and then just pass the final iterator into the insert_or_spawn_batch call here.

Reply:
I can't pass a query iterator outside of the system so I can't do this.